File: core/sdk/src/migrations/snapshotMigration0001.ts
import { Snapshot } from '@river-build/proto'
import { bin_equal } from '@river-build/dlog'

// Generic compactFunc function
function compactFunc<T>(elements: T[], keyFn: (element: T) => Uint8Array): T[] {
    if (elements.length === 0) {
        return elements
    }
    let j = 1
    for (let i = 1; i < elements.length; i++) {
        const key2 = keyFn(elements[i])
        const key1 = keyFn(elements[i - 1])

        if (!bin_equal(key2, key1)) {
            elements[j] = elements[i]
            j++
        }
    }

    return elements.slice(0, j)
}

// / nasty bug with the insert_sorted function, it was inserting an extra element at the end
// / every insert, we need to remove duplicates
export function snapshotMigration0001(snapshot: Snapshot): Snapshot {
    if (snapshot.members) {
        snapshot.members.joined = compactFunc(snapshot.members.joined, (m) => m.userAddress)
    }

    switch (snapshot.content?.case) {
        case 'spaceContent': {
            snapshot.content.value.channels = compactFunc(
                snapshot.content.value.channels,
                (c) => c.channelId,
            )
            break
        }
        case 'userContent': {
            snapshot.content.value.memberships = compactFunc(
                snapshot.content.value.memberships,
                (c) => c.streamId,
            )
            break
        }
        case 'userSettingsContent': {
            snapshot.content.value.fullyReadMarkers = compactFunc(
                snapshot.content.value.fullyReadMarkers,
                (c) => c.streamId,
            )
            break
        }
    }
    return snapshot
}

File: core/node/events/migrations/snapshot_migration_0001.go
package migrations

import (
	"bytes"
	"slices"

	. "github.com/river-build/river/core/node/protocol"
)

// nasty bug with the insert_sorted function, it was inserting an extra element at the end
// every insert, we need to remove duplicates
func snapshot_migration_0001(iSnapshot *Snapshot) *Snapshot {
	// gotta fix everywhere we used insertSorted, keep the first instance

	if iSnapshot.Members != nil {
		iSnapshot.Members.Joined = slices.CompactFunc(
			iSnapshot.Members.Joined,
			func(i, j *MemberPayload_Snapshot_Member) bool {
				return bytes.Equal(i.UserAddress, j.UserAddress)
			},
		)
	}

	switch snapshot := iSnapshot.Content.(type) {
	case *Snapshot_SpaceContent:
		if snapshot.SpaceContent != nil {
			snapshot.SpaceContent.Channels = slices.CompactFunc(
				snapshot.SpaceContent.Channels,
				func(i, j *SpacePayload_ChannelMetadata) bool {
					return bytes.Equal(i.ChannelId, j.ChannelId)
				},
			)
		}
	case *Snapshot_UserContent:
		if snapshot.UserContent != nil {
			snapshot.UserContent.Memberships = slices.CompactFunc(
				snapshot.UserContent.Memberships,
				func(i, j *UserPayload_UserMembership) bool {
					return bytes.Equal(i.StreamId, j.StreamId)
				},
			)
		}

	case *Snapshot_UserSettingsContent:
		if snapshot.UserSettingsContent != nil {
			snapshot.UserSettingsContent.FullyReadMarkers = slices.CompactFunc(
				snapshot.UserSettingsContent.FullyReadMarkers,
				func(i, j *UserSettingsPayload_FullyReadMarkers) bool {
					return bytes.Equal(i.StreamId, j.StreamId)
				},
			)
		}
	}

	return iSnapshot
}
